lib/core/src/Cardano/Wallet/Primitive/Migration/Planning.hs

```diff
@@ -0,0 +1,271 @@
+        case extendWith Supporter constraints (utxo, selection) of
+            Right (utxo', selection') ->
+                extendWithFreerider (utxo', selection')
+            Left ExtendSelectionAdaInsufficient ->
+                (utxo, selection)
+            Left ExtendSelectionEntriesExhausted ->
+                (utxo, selection)
+            Left ExtendSelectionFull ->
+                (utxo, selection)
+
+data ExtendSelectionError
+    = ExtendSelectionAdaInsufficient
+    | ExtendSelectionEntriesExhausted
+    | ExtendSelectionFull
+
+extendWith
+    :: TxSize s
+    => UTxOEntryCategory
+    -> TxConstraints s
+    -> (CategorizedUTxO i, Selection i s)
+    -> Either ExtendSelectionError (CategorizedUTxO i, Selection i s)
+extendWith category constraints (utxo, selection) =
+    case utxo `select` category of
+        Just (entry, utxo') ->
+            case Selection.extend constraints selection entry of
+                Right selection' ->
+                    Right (utxo', selection')
+                Left SelectionAdaInsufficient ->
+                    Left ExtendSelectionAdaInsufficient
+                Left SelectionFull {} ->
+                    Left ExtendSelectionFull
+        Nothing ->
+            Left ExtendSelectionEntriesExhausted
+
+select
+    :: CategorizedUTxO i
+    -> UTxOEntryCategory
+    -> Maybe ((i, TokenBundle), CategorizedUTxO i)
+select utxo = \case
+    Supporter -> selectSupporter
+    Freerider -> selectFreerider
+    Ignorable -> selectIgnorable
+  where
+    selectSupporter = case supporters utxo of
+        entry : remaining -> Just (entry, utxo {supporters = remaining})
+        [] -> Nothing
+    selectFreerider = case freeriders utxo of
+        entry : remaining -> Just (entry, utxo {freeriders = remaining})
+        [] ->  Nothing
+    selectIgnorable =
+        -- We never select an entry that should be ignored:
+        Nothing
+
+--------------------------------------------------------------------------------
+-- Categorization of UTxO entries
+--------------------------------------------------------------------------------
+
+data UTxOEntryCategory
+    = Supporter
+    -- ^ A coin or bundle that is capable of paying for its own marginal fee
+    -- and the base transaction fee.
+    | Freerider
+    -- ^ A coin or bundle that is not capable of paying for itself.
+    | Ignorable
+    -- ^ A coin that should not be added to a selection, because its value is
+    -- lower than the marginal fee for an input.
+    deriving (Eq, Show)
+
+data CategorizedUTxO i = CategorizedUTxO
+    { supporters :: ![(i, TokenBundle)]
+    , freeriders :: ![(i, TokenBundle)]
+    , ignorables :: ![(i, TokenBundle)]
+    }
+    deriving (Eq, Show)
+
+categorizeUTxO
+    :: TxSize s
+    => TxConstraints s
+    -> UTxO
+    -> CategorizedUTxO (TxIn, TxOut)
+categorizeUTxO constraints (UTxO u) = categorizeUTxOEntries constraints $
+    (\(i, o) -> ((i, o), view #tokens o)) <$> Map.toList u
```

Comment:
It seems `categorizeUTxO` and `uncategorizeUTxO` are both needed only in `createPlan`. 

How about moving the `entry <-> UTxO` conversion there as part of a `where`-clause?

Reply:
> It seems categorizeUTxO and uncategorizeUTxO are both needed only in createPlan.
> How about moving the entry <-> UTxO conversion there as part of a where-clause? It might be a bit clearer division of responsibility then.


Good idea! I've fixed this in d4819813bff15429724003d79aa489a77327a1fa.

Reply:
d4819813bff15429724003d79aa489a77327a1fa seems to have been dropped. Was this a mistake?